### PR TITLE
Implement transaction replacement fee bumping

### PR DESCRIPTION
**Pull request by shesek**
... plus some related fixups/changes.

### CODE REVIEW

**Comment by shesek** [comment]
The new tests don't pass because the feerate reported in the transaction list json is incorrect.

The last commit is a WIP attempt at fixing it, which does not yet work because the witness data is stripped away and not kept in the local db.

File: subprojects/gdk_rust/gdk_electrum/src/account.rs [hunk 1]
```diff
@@ -773,39 +789,72 @@ pub fn create_tx(
...
+    let mut change_addresses = HashMap::new();
+
+    // When a previous transaction is replaced, use it as a template for the new transaction
+    if let Some(ref prev_txitem) = request.previous_transaction {
+        if !request.addressees.is_empty() || send_all {
+            return Err(Error::InvalidReplacementRequest);
+        }
+
+        let prev_tx = BETransaction::from_hex(&prev_txitem.transaction, network.id())?;
+        let policy_asset = network.policy_asset.as_deref().unwrap_or("btc");
```
Comment (LeoComandini) [inline subprojects/gdk_rust/gdk_electrum/src/account.rs:805]: Elements/Liquid rbf support will not be exposed.
Do you prefer to leave the code to handle that anyway?

Comment (shesek) [reply subprojects/gdk_rust/gdk_electrum/src/account.rs:808]: I kept the code to handle it, but made it explicitly reject tx replacement for liquid networks in https://github.com/Blockstream/gdk/pull/119/commits/56db1ed00cd7ca8519a8dbb04ca7cb6a474330ed.

File: subprojects/gdk_rust/tests/integration.rs [hunk 2]
```diff
@@ -355,6 +355,52 @@ fn labels() {
...
+    // Replace it
+    let mut create_opt = CreateTransaction::default();
+    create_opt.subaccount = 1;
+    create_opt.previous_transaction = Some(txitem);
+    create_opt.fee_rate = Some(43000);
+    let tx = test_session.session.create_transaction(&mut create_opt).unwrap();
+    let signed_tx = test_session.session.sign_transaction(&tx).unwrap();
+    let txid = test_session.session.broadcast_transaction(&signed_tx.hex).unwrap();
+    test_session.wait_account_tx(1, &txid);
+    let txitem = test_session.get_tx_from_list(1, &txid);
```
Comment (LeoComandini) [inline subprojects/gdk_rust/tests/integration.rs:399]: Is the replaced tx still in the tx list?
In both cases, we should add few lines to test the intended behavior

File: subprojects/gdk_rust/tests/integration.rs [hunk 3]
```diff
@@ -355,6 +355,52 @@ fn labels() {
...
+    assert_eq!(txitem.fee_rate / 1000, 43);
+    assert_eq!(txitem.memo, "poz qux");
```
Comment (LeoComandini) [inline subprojects/gdk_rust/tests/integration.rs:401]:
```suggestion
    assert_eq!(txitem.memo, "poz qux");

    test_session.stop();
```

File: subprojects/gdk_rust/gdk_electrum/src/error.rs [hunk 4]
```diff
@@ -59,6 +60,9 @@ impl Display for Error {
             Error::AccountGapsDisallowed => {
                 write!(f, "cannot create a new subaccount while the last one is unused")
             }
+            Error::InvalidReplacementRequest => {
+                write!(f, "invalid replacement request fields")
+            }
```
Comment (LeoComandini) [inline subprojects/gdk_rust/gdk_electrum/src/error.rs:65]:
```suggestion
            Error::InvalidReplacementRequest => write!(f, "invalid replacement request fields"),
```

File: subprojects/gdk_rust/gdk_electrum/src/lib.rs [hunk 5]
```diff
@@ -287,11 +286,10 @@ fn make_txlist_item(tx: &TransactionMeta) -> TxListItem {
         type_,
         memo: tx.create_transaction.as_ref().and_then(|c| c.memo.clone()).unwrap_or("".to_string()),
         txhash: tx.txid.clone(),
-        transaction_size: len,
         transaction: tx.hex.clone(), // FIXME
         satoshi: tx.satoshi.clone(),
         rbf_optin: tx.rbf_optin, // TODO: TransactionMeta -> TxListItem rbf_optin
-        cap_cpfp: false,         // TODO: TransactionMeta -> TxListItem cap_cpfp
+        can_cpfp: false,         // TODO: TransactionMeta -> TxListItem can_cpfp
         can_rbf: false,          // TODO: TransactionMeta -> TxListItem can_rbf
```
Comment (LeoComandini) [inline subprojects/gdk_rust/gdk_electrum/src/lib.rs:293]: This should be true if tx is not confirmed and tx.rbf_optin

File: subprojects/gdk_rust/gdk_common/src/model.rs [hunk 6]
```diff
@@ -312,6 +326,20 @@ pub struct TxListItem {
     pub transaction_weight: usize,
 }
 
+// Negative (sent) amounts are expceted to be provided as positive numbers.
```
Comment (jgriffiths) [inline subprojects/gdk_rust/gdk_common/src/model.rs:329]: nit: `expected`

**Comment by LeoComandini** [comment]
In src/ga_rust.cpp#L417, something like this is needed:

```
         auto addressees_p = result.find("addressees");
-        for (auto& addressee : *addressees_p) {
+        if (addressees_p != result.end()) {
+            for (auto& addressee : *addressees_p) {
...
```

And also `CreateTransaction.addressees` should become an `Option`, otherwise caller is forced to specified that as well when bumping a transaction.

File: subprojects/gdk_rust/tests/integration.rs (hunk 2 above)
Comment (shesek) [reply subprojects/gdk_rust/tests/integration.rs]: To my understanding it should not be in the list, the same as how gdk-cpp behaves.

But there was a bug that prevented this from working, fixed in https://github.com/Blockstream/gdk/pull/119/commits/278ac0996ae879652aee29d9dbcb0f69d0c7e14b and added a test in https://github.com/Blockstream/gdk/pull/119/commits/dcd661c2c49235930c987018aa797337b6d3b24c.

File: subprojects/gdk_rust/gdk_common/src/model.rs (hunk 6 above)
Comment (shesek) [reply subprojects/gdk_rust/gdk_common/src/model.rs]: Fixed in https://github.com/Blockstream/gdk/pull/119/commits/cd2e9cd1ad35823b9f6b7143835c0165597704b8.

File: subprojects/gdk_rust/tests/integration.rs (hunk 3 above)
Comment (shesek) [reply subprojects/gdk_rust/tests/integration.rs:403]: Added in https://github.com/Blockstream/gdk/pull/119/commits/60c7dd3bb5bbcc678e8d265a0d32a9bcb56efc4f.

File: subprojects/gdk_rust/gdk_electrum/src/lib.rs (hunk 5 above)
Comment (shesek) [reply subprojects/gdk_rust/gdk_electrum/src/lib.rs]: Addressed in https://github.com/Blockstream/gdk/pull/119/commits/b0796b44e05c7f30514e6390a50fa7c2c150dfcd.

File: subprojects/gdk_rust/tests/integration.rs [hunk 7]
```diff
@@ -355,6 +357,78 @@ fn labels() {
...
+    let tx = test_session.session.create_transaction(&mut create_opt).unwrap();
+    let signed_tx = test_session.session.sign_transaction(&tx).unwrap();
+    let txid2 = test_session.session.broadcast_transaction(&signed_tx.hex).unwrap();
+    test_session.wait_account_tx(1, &txid2);
+    let txitem = test_session.get_tx_from_list(1, &txid2);
+    assert_eq!(txitem.fee_rate / 1000, 43);
+    assert_eq!(txitem.memo, "poz qux");
+
+    // The old transaction should be gone (after the next sync with the server)
+    std::thread::sleep(std::time::Duration::from_secs(3));
```
Comment (LeoComandini) [inline subprojects/gdk_rust/tests/integration.rs:406]: Sometimes this is not enough, I got failures both in CI and locally.
Can you do something similar to `get_tx_from_list`? i.e. loop until the tx is not in the list anymore for 1 minute at most.
Or anything that you think is more appropriate

**Comment by LeoComandini** [comment]
ACK 510b56d0, code review, run CI tests and performed manual tests locally, gitlab CI passed, will add cpp change required separately.

Thanks @shesek !!

**Pull request merged**